File: 0844 - Backspace String Compare/solution.py
class Solution:
    def backspaceCompare(self, s1: str, s2: str) -> bool:
        i1, i2, n1, n2 = len(s1)-1, len(s2)-1, 0, 0
        while True:
            while i1>=0 and (n1 or s1[i1]=='#'):
                n1 += 1 if s1[i1] == '#' else -1
                i1 -= 1
            while i2>=0 and (n2 or s2[i2]=='#'):
                n2 += 1 if s2[i2] == '#' else -1
                i2 -= 1
            if i1<0 or i2<0 or s1[i1]!=s2[i2]:
                return i1 == i2 == -1
            i1, i2 = i1-1, i2-1
        return true

File: 0844 - Backspace String Compare/solution.cpp
class Solution {
private:
    int enter(string& s){
        int idx = 0;
        for(int i = 0; i < s.size(); i++){
            if(s[i] == '#') idx = max(0, idx-1);
            else s[idx++] = s[i];
        }
        return idx;
    }
public:
    bool backspaceCompare(string s1, string s2) {
        int n1 = enter(s1), n2 = enter(s2);
        if(n1 != n2) return false;
        for(int i = 0; i < n1; i++)
            if(s1[i] != s2[i]) return false;
        return true;
    }
};